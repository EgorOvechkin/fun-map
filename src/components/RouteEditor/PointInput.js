import React, { PureComponent } from 'react';
import './styles/PointInput.css';

class RouteEditor extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      pointTitle: '',
      isValid: true
    };

    this.handleChange = this.handleChange.bind(this);
    this.keyPress = this.keyPress.bind(this);
    this.onBlur = this.onBlur.bind(this);
  }

  render() {
    return (
      <input
        className={this.state.isValid ? 'point-input' : 'point-input error'}
        type="text"
        value={this.state.pointTitle}s
        onBlur={this.onBlur}
        onChange={this.handleChange}
        onKeyDown={this.keyPress}
        placeholder={'Новая точка маршрута'}
      />
    );
  }

  onBlur() {
    if (!this.state.isValid) {
      this.setState({isValid: true});
    }
  }

  handleChange(event) {
    const newValue = event.target.value;

    this.setState({pointTitle: newValue});

    if (!this.state.isValid) {
      this.setState({isValid: newValue.trim() !== ''});
    }
  }

  keyPress(event) {
    if (event.key !== 'Enter') {
      return;
    }

    if (!this.state.pointTitle.trim()) {
      this.setState({isValid: false, pointTitle: ''});
      return;
    }

    this.props.addPoint(this.state.pointTitle.trim());
    this.setState({pointTitle: ''});
  }
}

export default RouteEditor;
