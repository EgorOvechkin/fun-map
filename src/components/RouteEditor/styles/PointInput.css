.point-input {
  font-size: 1.5em;
  height: 2em;
  box-sizing: border-box;
  width: 100%;
}

.point-input::placeholder {
  color: #bababa;
}
