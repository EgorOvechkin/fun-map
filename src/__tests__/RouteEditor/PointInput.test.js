import React from 'react';
import PointInput from '../../components/RouteEditor/PointInput';
import {shallow} from 'enzyme';

let wrapper;
beforeEach(() => wrapper = shallow(<PointInput />));

it('renders valid point input', () => {
  wrapper.setState({isValid: true});
  expect(wrapper).toMatchSnapshot();
});

it('renders not valid point input', () => {
  wrapper.setState({isValid: false});
  expect(wrapper).toMatchSnapshot();
});

describe('onBlur', () => {
  it('changes isValid from false to true', () => {
    wrapper.simulate('focus');
    wrapper.setState({isValid: false});
    wrapper.simulate('blur');
    expect(wrapper.state('isValid')).toBeTruthy();
  });

  it(`doesn't change isValid when it is true`, () => {
    wrapper.setState({isValid: true});
    wrapper.simulate('focus');
    wrapper.simulate('blur');
    expect(wrapper.state('isValid')).toBeTruthy();
  });
});

describe('handleChange', () => {
  describe('when value is not empty', () => {
    it('saves input value', () => {
      const newValue = 'point title';

      wrapper.simulate('change', {target: {value: newValue}});
      expect(wrapper.state('pointTitle')).toBe(newValue);
    });
  });

  describe('when isValid is false and new value is empty string', () => {
    it(`sets isValid as false and pointTitle as ''`, () => {
      const newValue = '   ';
      wrapper.setState({isValid: false, pointTitle: 'old point title'});

      wrapper.simulate('change', {target: {value: newValue}});
      expect(wrapper.state('isValid')).toBeFalsy();
      expect(wrapper.state('pointTitle')).toBe(newValue);
    });
  });

  describe('when isValid is false and new value is not empty string', () => {
    it('sets isValid as true and saves new value', () => {
      const newValue = 'new point title';
      wrapper.setState({isValid: false, pointTitle: 'old point title'});

      wrapper.simulate('change', {target: {value: newValue}});
      expect(wrapper.state('isValid')).toBeTruthy();
      expect(wrapper.state('pointTitle')).toBe(newValue);
    });
  });

  describe('when isValid is true', () => {
    it('saves new value', () => {
      const newValue = 'new point title';
      wrapper.setState({isValid: true, pointTitle: 'old point title'});

      wrapper.simulate('change', {target: {value: newValue}});
      expect(wrapper.state('isValid')).toBeTruthy();
      expect(wrapper.state('pointTitle')).toBe(newValue);
    });
  });
});

describe('keyPress', () => {
  describe('when key is not Enter', () => {
    it('does nothing', () => {
      const event = {key: 'Up'};
      const oldState = {isValid: true, pointTitle: 'point'};

      wrapper.setState(oldState);
      wrapper.simulate('keydown', event);
      expect(wrapper.state()).toMatchObject(oldState);
    });
  });

  describe('when key is Enter', () => {
    describe('when pointTitle is empty', () => {
      it('sets isValid as flase', () => {
        wrapper.setState({pointTitle: '', isValid: true});
        wrapper.simulate('keydown', {key: 'Enter'});

        expect(wrapper.state('isValid')).toBeFalsy();
      });

    });
    describe('when pointTitle is not empty', () => {
      it('calls addPoint and sets empty point title', () => {
        const addPoint = jest.fn();
        const event= {key: 'Enter'};
        const pointTitle = 'Point';
        wrapper.setState({pointTitle: pointTitle, isValid: true});
        wrapper.setProps({addPoint: addPoint});
        wrapper.simulate('keydown', event);

        expect(addPoint.mock.calls.length).toBe(1);
        expect(addPoint.mock.calls[0][0]).toBe(pointTitle);
        expect(wrapper.state('pointTitle')).toBe('');
      });
    });
  });
});
