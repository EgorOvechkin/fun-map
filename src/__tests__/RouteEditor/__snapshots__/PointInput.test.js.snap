// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders not valid point input 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RouteEditor />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "className": "point-input error",
      "onBlur": [Function],
      "onChange": [Function],
      "onKeyDown": [Function],
      "placeholder": "Новая точка маршрута",
      "s": true,
      "type": "text",
      "value": "",
    },
    "ref": null,
    "rendered": null,
    "type": "input",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": "point-input error",
        "onBlur": [Function],
        "onChange": [Function],
        "onKeyDown": [Function],
        "placeholder": "Новая точка маршрута",
        "s": true,
        "type": "text",
        "value": "",
      },
      "ref": null,
      "rendered": null,
      "type": "input",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`renders valid point input 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RouteEditor />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "className": "point-input",
      "onBlur": [Function],
      "onChange": [Function],
      "onKeyDown": [Function],
      "placeholder": "Новая точка маршрута",
      "s": true,
      "type": "text",
      "value": "",
    },
    "ref": null,
    "rendered": null,
    "type": "input",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": "point-input",
        "onBlur": [Function],
        "onChange": [Function],
        "onKeyDown": [Function],
        "placeholder": "Новая точка маршрута",
        "s": true,
        "type": "text",
        "value": "",
      },
      "ref": null,
      "rendered": null,
      "type": "input",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
